CLI command to clear the Thaw cache

Would be useful to be able to clear the Thaw cache under `~/.thaw` with a CLI command: `thaw clean`

Hi,
Thanks for the instruction. :smile: 

I tried to run the application with the following command in Ubuntu
`./gradlew :cli:run --args="--root-folder='../example/demo' --output='../example/demo/demo.pdf'"`

But I am getting the following erro-
```bash
########################
### Thaw Project CLI ###
########################

### Configuration ###
Charset: 'UTF-8'

### Processing ###
Searching for Thaw files within folder at '/home/badal/Documents/codeProject/thaw/cli/../example/demo'...

Processing Thaw info file 'demo.tdi'...
Processing Thaw text file 'demo.tdt'...
Processing Thaw style file 'demo.tds'...
Oct. 14, 2020 10:57:52 P.M. org.apache.fontbox.ttf.CmapSubtable processSubtype14
WARNING: Format 14 cmap table is not supported and will be ignored
An exception occurred while trying to export the resulting document.
The exception message is: 'de.be.thaw.typeset.exception.TypeSettingException: Could not convert the document into the Knuth-Plass algorithm format'
de.be.thaw.export.exception.ExportException: de.be.thaw.typeset.exception.TypeSettingException: Could not convert the document into the Knuth-Plass algorithm format
        at de.be.thaw.export.pdf.PdfExporter.export(PdfExporter.java:109)
        at de.be.thaw.cli.CLI.call(CLI.java:236)
        at de.be.thaw.cli.CLI.call(CLI.java:36)
        at picocli.CommandLine.executeUserObject(CommandLine.java:1853)
        at picocli.CommandLine.access$1100(CommandLine.java:145)
        at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2255)
        at picocli.CommandLine$RunLast.handle(CommandLine.java:2249)
        at picocli.CommandLine$RunLast.handle(CommandLine.java:2213)
        at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2080)
        at picocli.CommandLine.execute(CommandLine.java:1978)
        at de.be.thaw.cli.CLI.main(CLI.java:69)
Caused by: de.be.thaw.typeset.exception.TypeSettingException: Could not convert the document into the Knuth-Plass algorithm format
        at de.be.thaw.typeset.knuthplass.KnuthPlassTypeSetter.convertToParagraphs(KnuthPlassTypeSetter.java:177)
        at de.be.thaw.typeset.knuthplass.KnuthPlassTypeSetter.typesetWithConfig(KnuthPlassTypeSetter.java:89)
        at de.be.thaw.typeset.knuthplass.KnuthPlassTypeSetter.typeset(KnuthPlassTypeSetter.java:77)
        at de.be.thaw.export.pdf.PdfExporter.typeset(PdfExporter.java:219)
        at de.be.thaw.export.pdf.PdfExporter.export(PdfExporter.java:107)
        ... 10 more
Caused by: de.be.thaw.core.document.convert.exception.DocumentConversionException: de.be.thaw.export.exception.ExportException: Could not find font family 'Arial' in OS installed fonts. If on Windows make sure that the fonts were installed for all users for Thaw to be able to pick it up.
        at de.be.thaw.typeset.knuthplass.converter.context.ConversionContext.appendTextToParagraph(ConversionContext.java:198)
        at de.be.thaw.typeset.knuthplass.converter.KnuthPlassConverter.initializeTextualNode(KnuthPlassConverter.java:153)
        at de.be.thaw.typeset.knuthplass.converter.KnuthPlassConverter.initializeForNode(KnuthPlassConverter.java:118)
        at de.be.thaw.typeset.knuthplass.converter.KnuthPlassConverter.initializeForNode(KnuthPlassConverter.java:126)
        at de.be.thaw.typeset.knuthplass.converter.KnuthPlassConverter.initializeForNode(KnuthPlassConverter.java:126)
        at de.be.thaw.typeset.knuthplass.converter.KnuthPlassConverter.convert(KnuthPlassConverter.java:100)
        at de.be.thaw.typeset.knuthplass.KnuthPlassTypeSetter.convertToParagraphs(KnuthPlassTypeSetter.java:175)
        ... 14 more
Caused by: de.be.thaw.export.exception.ExportException: Could not find font family 'Arial' in OS installed fonts. If on Windows make sure that the fonts were installed for all users for Thaw to be able to pick it up.
        at de.be.thaw.export.pdf.util.ExportContext.lambda$getFontVariantLocatorForStyleValue$0(ExportContext.java:322)
        at java.base/java.util.Optional.orElseThrow(Optional.java:401)
        at de.be.thaw.export.pdf.util.ExportContext.getFontVariantLocatorForStyleValue(ExportContext.java:322)
        at de.be.thaw.export.pdf.util.ExportContext.getFontForNode(ExportContext.java:363)
        at de.be.thaw.export.pdf.PdfExporter$3.getSpaceWidth(PdfExporter.java:270)
        at de.be.thaw.typeset.knuthplass.converter.context.ConversionContext.appendTextToParagraph(ConversionContext.java:193)
        ... 20 more

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':cli:run'.
> Process 'command '/home/badal/.jabba/jdk/openjdk@1.14.0/bin/java'' finished with non-zero exit value 4
```

Not sure if I need to do any additional set up.

I changed the `demo.tds` file with `Ubuntu` font like below -
 ```bash
	font-family: Ubuntu:url(/usr/share/fonts/truetype/ubuntu/);
	inline-code-font-family: Ubuntu:url(usr/share/fonts/truetype/ubuntu/);
```
But I am getting the following error- 

```bash
### Processing ###
Searching for Thaw files within folder at '/home/badal/Documents/codeProject/thaw/cli/../example/demo'...

Processing Thaw info file 'demo.tdi'...
Processing Thaw text file 'demo.tdt'...
Processing Thaw style file 'demo.tds'...
An exception occurred while trying to parse the provided style file at '/home/badal/Documents/codeProject/thaw/cli/../example/demo/demo.tds'.
The exception message is: 'de.be.thaw.style.parser.value.exception.StyleValueParseException: Could not resolve font file at 'usr/share/fonts/truetype/ubuntu/''

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':cli:run'.
> Process 'command '/home/badal/.jabba/jdk/openjdk@1.14.0/bin/java'' finished with non-zero exit value 8

```

My directory has the following contents-

![ubuntufont](https://user-images.githubusercontent.com/19233335/96480030-ace9b300-1207-11eb-9f3a-1211b535fbeb.png)

It says, it can't resolve the font file. But the files are in the folder. Any idea what can be wrong?

I will also switch to Windows in the meantime and finish #62.

--- a/THIRD_PARTY_NOTICE.md
+++ b/THIRD_PARTY_NOTICE.md
@@ -86,7 +86,16 @@ You're able to obtain a copy of the license from https://github.com/michel-kraem
 
 ## Apache Commons IO
 
-This [library](https://github.com/apache/commons-io) is Licensed under the
-Apache License, Version 2.0 (the "License").you may not use this file except in
-compliance with the License. You may obtain a copy of the License at
+This [library](https://github.com/apache/commons-io) is licensed under the
+Apache License, Version 2.0 (the "License"). You may not use this file except in
+compliance with the license. You may obtain a copy of it at
 [http://www.apache.org/licenses/LICENSE-2.0](http://www.apache.org/licenses/LICENSE-2.0).
+
+Additionally the attribution notices of Apache Commons IO are included:
+```
+Apache Commons IO
+Copyright 2002-2020 The Apache Software Foundation
+
+This product includes software developed at
+The Apache Software Foundation (https://www.apache.org/).
+```

--- a/THIRD_PARTY_NOTICE.md
+++ b/THIRD_PARTY_NOTICE.md
@@ -83,3 +83,10 @@ It is distributed under the Apache-2.0 License which can be obtained from https:
 
 citeproc-java is used (https://github.com/michel-kraemer/citeproc-java) which is distributed unter the Apache-2.0 License.
 You're able to obtain a copy of the license from https://github.com/michel-kraemer/citeproc-java/blob/master/LICENSE.txt or http://www.apache.org/licenses/LICENSE-2.0.
+
+## Apache Commons IO
+
+This [library](https://github.com/apache/commons-io) is Licensed under the
+Apache License, Version 2.0 (the "License").you may not use this file except in
+compliance with the License. You may obtain a copy of the License at
+[http://www.apache.org/licenses/LICENSE-2.0](http://www.apache.org/licenses/LICENSE-2.0).

--- a/cli/src/main/java/de/be/thaw/cli/CLI.java
+++ b/cli/src/main/java/de/be/thaw/cli/CLI.java
@@ -19,6 +19,7 @@ import de.be.thaw.style.parser.exception.StyleModelParseException;
 import de.be.thaw.text.model.TextModel;
 import de.be.thaw.text.parser.TextParser;
 import de.be.thaw.text.parser.exception.ParseException;
+import de.be.thaw.util.cache.CacheUtil;
 import picocli.CommandLine;
 
 import java.io.BufferedReader;
@@ -29,6 +30,7 @@ import java.nio.charset.Charset;
 import java.nio.file.Path;
 import java.util.Locale;
 import java.util.concurrent.Callable;
+import java.io.IOException;
 
 /**
  * CLI entry point for Thaw.
@@ -59,6 +61,21 @@ public class CLI implements Callable<Integer> {
     @CommandLine.Option(names = {"-c", "--charset"}, description = "Name of the charset the files to process are encoded in. " +
             "If not specified the systems default charset will be used.")
     private String charsetName;
+
+    /**
+     * Subcommand to clean the cache folder
+     */
+    @CommandLine.Command(name="clean", description="Empty the root cache folder.")
+    int clean(){
+        try{
+            CacheUtil.cleanCacheRootDir();
+        }
+        catch(IOException e){
+            System.err.println("An exception has occurred. Root cache clean is not successful. Try again later.");
+            return ErrorResult.ROOT_CACHE_CLEANING_ERROR.getCode();
+        }
+        return ErrorResult.OK.getCode();
+    }
 
     /**
      * Entry point of the CLI application.

--- a/cli/src/main/java/de/be/thaw/cli/ErrorResult.java
+++ b/cli/src/main/java/de/be/thaw/cli/ErrorResult.java
@@ -15,7 +15,8 @@ public enum ErrorResult {
     MORE_THAN_ONE_STYLE_FILE(7),
     STYLE_FILE_PARSING_ERROR(8),
     MORE_THAN_ONE_SOURCE_FILE(9),
-    SOURCE_FILE_PARSING_ERROR(10);
+    SOURCE_FILE_PARSING_ERROR(10),
+    ROOT_CACHE_CLEANING_ERROR(11);
 
     /**
      * Code of the error.

--- a/util/build.gradle
+++ b/util/build.gradle
@@ -1,3 +1,4 @@
 dependencies {
     // Additional dependencies go here
+    compile 'commons-io:commons-io:2.8.0'
 }

--- a/util/src/main/java/de/be/thaw/util/cache/CacheUtil.java
+++ b/util/src/main/java/de/be/thaw/util/cache/CacheUtil.java
@@ -1,6 +1,7 @@
 package de.be.thaw.util.cache;
 
 import de.be.thaw.util.cache.exception.CouldNotGetProjectCacheDirectoryException;
+import org.apache.commons.io.FileUtils;
 
 import java.io.ByteArrayInputStream;
 import java.io.File;
@@ -36,6 +37,22 @@ public class CacheUtil {
         }
 
         return dir;
+    }
+
+    /**
+     * Clean the root caching directory.
+     * Tries 3 times in case of IOException.
+     *
+     * @throws IOException  in case directory access failed or other errors related to IO
+     */
+    public static void cleanCacheRootDir() throws IOException {
+        try{
+            FileUtils.cleanDirectory(getCacheRootDir());
+        }
+        catch(IOException e){
+            throw e;
+        }
+        System.out.println("Root cache cleaned successfully");
     }
 
     /**

--- a/util/src/main/java/module-info.java
+++ b/util/src/main/java/module-info.java
@@ -1,4 +1,5 @@
 module de.be.thaw.util {
+    requires org.apache.commons.io;
     exports de.be.thaw.util;
     exports de.be.thaw.util.os.exception;
     exports de.be.thaw.util.os;

